### PR TITLE
Remove cluster level list/watch for configmaps, serviceaccounts, secrets

### LINKED ISSUE #366: service_account_credentials implementation requires cluster level get/list/watch of ServiceAccounts and Secrets
/kind bug

**What steps did you take and what happened:**

As I work on https://github.com/cliveseldon/seldon-core/pull/1/, I realized this code path is copied from/shared with kfserving, and I would like to report it as a "bug".

**What did you expect to happen:**

The controller does not require cluster wide GET/LIST/WATCH of ServiceAccounts and Secrets, only needed in the namespace it needs to.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

A patch is proposed in https://github.com/cliveseldon/seldon-core/pull/1/

**Environment:**

- Istio Version:
- Knative Version:
- KFServing Version:
- Kubeflow version:
- Minikube version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### LINKED ISSUE #3467: Controller out-of-memory kill when cluster has too many secrets
/kind bug

**What steps did you take and what happened:**
Given a cluster with a high count of Secrets (in the order of about 8k secrets), when and ISVC is reconciled the KServe controller tries to load into memory all Secrets in the cluster. Because the high count of Secrets, the KServe controller will reach its memory limit and the cluster will kill the pod. The cluster may re-create the pod but, given unchanged conditions, the pod will be killed continuously.

**What did you expect to happen:**
KServe controller should work fine even if the cluster has a high count of Secrets.

**What's the InferenceService yaml:**
Not relevant. Any ISVC will lead to the crash.

**Anything else you would like to add:**
On a quick code review, looks like KServe catches the secrets when using an informer/watch. However, KServe may not need to watch for changes in Secrets and, apparently, Secrets are always queried by name. If this is true, it may be better to query directly the API Server rather than via a cache of the Secrets, preventing the need of an Informer/watch.

**Environment:**
- KServe Version: 0.11.2 - should also happen in  `master`

### PR DESCRIPTION
**Pull request by sivanantha321**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3467
Fixes #366

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Remove cluster level list/watch for configmaps, serviceaccounts, secrets
```

### CODE REVIEW

**Comment by sivanantha321** [comment]
@israel-hdez Can you help test whether this fixes the out-of-memory issue ?

**Review by terrytangyuan — COMMENTED** [review]
Thank you for fixing this!

**Comment by israel-hdez** [comment]
> @israel-hdez Can you help test whether this fixes the out-of-memory issue ?

I'll test on my Wednesday.

**Comment by terrytangyuan** [comment]
@sivanantha321 Could you point me to where you removed the list calls for secrets? Perhaps I missed it somewhere but I only see the removal from manifests.

File: pkg/controller/v1beta1/inferenceservice/components/explainer.go [hunk 1]
```diff
@@ -45,17 +46,19 @@ var _ Component = &Explainer{}
...
 	scheme                 *runtime.Scheme
 	inferenceServiceConfig *v1beta1.InferenceServicesConfig
 	credentialBuilder      *credentials.CredentialBuilder //nolint: unused
 	deploymentMode         constants.DeploymentModeType
 	Log                    logr.Logger
 }
 
-func NewExplainer(client client.Client, scheme *runtime.Scheme, inferenceServiceConfig *v1beta1.InferenceServicesConfig,
-	deploymentMode constants.DeploymentModeType) Component {
+func NewExplainer(client client.Client, clientest kubernetes.Interface, scheme *runtime.Scheme,
```
Comment (terrytangyuan) [inline pkg/controller/v1beta1/inferenceservice/components/explainer.go]: clientest -> clientset

File: pkg/controller/v1beta1/inferenceservice/reconcilers/cabundleconfigmap/cabundle_configmap_reconciler.go [hunk 2]
```diff
@@ -91,10 +90,9 @@ func (c *CaBundleConfigMapReconciler) getCabundleConfigMapForUserNS(caBundleName
 
 	// Check if cabundle configmap exist & the cabundle.crt exist in the data in controller namespace
 	// If it does not exist, return error
-	caBundleConfigMap := &corev1.ConfigMap{}
-	if err := c.client.Get(context.TODO(),
-		types.NamespacedName{Name: caBundleNameInConfig, Namespace: kserveNamespace}, caBundleConfigMap); err == nil {
+	caBundleConfigMap, err := c.clientset.CoreV1().ConfigMaps(kserveNamespace).Get(context.TODO(), caBundleNameInConfig, metav1.GetOptions{})
 
+	if err == nil {
```
Comment (terrytangyuan) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/cabundleconfigmap/cabundle_configmap_reconciler.go:96]: The else block's error message might be misleading. It should be "cannot get the configmap from the cluster"

**Comment by israel-hdez** [comment]
> @israel-hdez Can you help test whether this fixes the out-of-memory issue ?

@sivanantha321 I'm still replicating the setup. I tried today, and my cluster died while on it.
I'll re-try on my Friday.

**Comment by sivanantha321** [comment]
> @sivanantha321 Could you point me to where you removed the list calls for secrets? Perhaps I missed it somewhere but I only see the removal from manifests.

We are not using list calls in the code base.

**Comment by israel-hdez** [comment]
I've just replicated the OOMKill with `master` branch to ensure I have the needed setup.

I used a script from @skonto available in this gist: https://gist.github.com/skonto/188b2c45c5af449629caaa69a190392c, with a minor modification: edit line 3 to `NC=64`, because that was enough in my environment. The script would create about 8192 secrets, to a total of about 9528 secrets in my env.

I'm now trying the code changes in this PR.

**Comment by israel-hdez** [comment]
I confirm the kserve-controller pod no longer crashes with OOMKill.
I'll do code review soon.

Notice that I only tried around _Secrets_ which is what I reported in #3467. I didn't try the other resources.

**Comment by terrytangyuan** [comment]
> We are not using list calls in the code base.

I meant "list/watch for secrets" as described in the PR title but thanks I got it now.

File: config/rbac/role.yaml [hunk 3]
```diff
@@ -48,9 +48,7 @@ rules:
   verbs:
   - create
   - get
-  - list
```
Comment (israel-hdez) [inline config/rbac/role.yaml:51]: Should the helm charts be updated?
I don't remember if they are updated at release.

File: pkg/apis/serving/v1beta1/inference_service_defaults.go [hunk 4]
```diff
@@ -18,15 +18,15 @@ package v1beta1
 
 import (
 	"fmt"
+	"k8s.io/client-go/kubernetes"
```
Comment (israel-hdez) [inline pkg/apis/serving/v1beta1/inference_service_defaults.go]: reorganize the imports?

File: pkg/controller/v1alpha1/trainedmodel/controller.go [hunk 5]
```diff
@@ -18,24 +18,19 @@ limitations under the License.
...
+// +kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;update
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;create;update;patch;delete
 // +kubebuilder:rbac:groups=core,resources=namespaces,verbs=get;list;watch
 // +kubebuilder:rbac:groups=core,resources=events,verbs=get;list;watch;create;update;patch;delete
 package trainedmodel
 
 import (
 	"context"
 	"fmt"
+	"k8s.io/client-go/kubernetes"
```
Comment (israel-hdez) [inline pkg/controller/v1alpha1/trainedmodel/controller.go]: reorganize?

File: pkg/controller/v1alpha1/trainedmodel/reconcilers/modelconfig/modelconfig_reconciler.go [hunk 6]
```diff
@@ -19,14 +19,14 @@ package modelconfig
 import (
 	"context"
 	"fmt"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (israel-hdez) [inline pkg/controller/v1alpha1/trainedmodel/reconcilers/modelconfig/modelconfig_reconciler.go:23]: reorganize?

**Review by israel-hdez — COMMENTED** [review]
Leaving a few minor comments.

File: pkg/controller/v1beta1/inferenceservice/components/explainer.go (hunk 1 above)
Comment (sivanantha321) [reply pkg/controller/v1beta1/inferenceservice/components/explainer.go]: Fixed

File: pkg/controller/v1beta1/inferenceservice/reconcilers/cabundleconfigmap/cabundle_configmap_reconciler.go (hunk 2 above)
Comment (sivanantha321) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/cabundleconfigmap/cabundle_configmap_reconciler.go:96]: Done

File: config/rbac/role.yaml (hunk 3 above)
Comment (sivanantha321) [reply config/rbac/role.yaml:51]: Updated

File: pkg/apis/serving/v1beta1/inference_service_defaults.go (hunk 4 above)
Comment (sivanantha321) [reply pkg/apis/serving/v1beta1/inference_service_defaults.go]: Done

File: pkg/controller/v1alpha1/trainedmodel/controller.go (hunk 5 above)
Comment (sivanantha321) [reply pkg/controller/v1alpha1/trainedmodel/controller.go]: Done

File: pkg/controller/v1alpha1/trainedmodel/reconcilers/modelconfig/modelconfig_reconciler.go (hunk 6 above)
Comment (sivanantha321) [reply pkg/controller/v1alpha1/trainedmodel/reconcilers/modelconfig/modelconfig_reconciler.go:23]: Done

File: pkg/controller/v1beta1/inferenceservice/components/explainer.go [hunk 7]
```diff
@@ -19,14 +19,9 @@ package components
 import (
 	"context"
 	"fmt"
+	"k8s.io/client-go/kubernetes"
```
Comment (israel-hdez) [inline pkg/controller/v1beta1/inferenceservice/components/explainer.go:22]: ?

File: pkg/controller/v1beta1/inferenceservice/reconcilers/cabundleconfigmap/cabundle_configmap_reconciler.go [hunk 8]
```diff
@@ -20,42 +20,42 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
-
-	kservev1beta1 "github.com/kserve/kserve/pkg/apis/serving/v1beta1"
-	"github.com/kserve/kserve/pkg/constants"
-	"github.com/kserve/kserve/pkg/webhook/admission/pod"
+	"github.com/pkg/errors"
```
Comment (israel-hdez) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/cabundleconfigmap/cabundle_configmap_reconciler.go:23]: ?

**Comment by israel-hdez** [comment]
Per request in [prow comment](https://github.com/kserve/kserve/pull/3469#issuecomment-1959965577):

/assign @rachitchauhan43

**Review by israel-hdez — COMMENTED** [review]
/lgtm

**Review by terrytangyuan — APPROVED** [review]
/lgtm

**Comment by israel-hdez** [comment]
I pinged Rachit previously.
But now I'm also pinging @yuzisun in case Rachit is not available.

File: cmd/manager/main.go [hunk 9]
```diff
@@ -111,6 +110,11 @@ func main() {
 		setupLog.Error(err, "unable to set up client config")
 		os.Exit(1)
 	}
+	clientSet, err := kubernetes.NewForConfig(cfg)
```
Comment (rachitchauhan43) [inline cmd/manager/main.go:116]: Can we put inline comment on for what this is for just like other objects we are creating ?

**Review by rachitchauhan43 — CHANGES_REQUESTED** [review]
A small change requested. Rest LGTM.

**Comment by rachitchauhan43** [comment]
Looks like lots of `imports` reorganization happened in this PR. @sivanantha321 : Possible to add as an action itself that fails if imports are not as per standard in `go` files ?

**Comment by rachitchauhan43** [comment]
Also, @israel-hdez , if we have reproducibility, what do you think of adding a test case to make sure regression always catches such issues  ?

**Comment by israel-hdez** [comment]
> Also, @israel-hdez , if we have reproducibility, what do you think of adding a test case to make sure regression always catches such issues ?

@rachitchauhan43 I do know how to reproduce, but IMO that should be solved under its own ticket, because we would need to find a way to replicate within CI -- I crashed my dev cluster a few times before I could replicate successfully, so what I did on my device may not be suitable for GitHub Actions. And, of course, we would need to ensure all relevant code paths are covered.

**Comment by israel-hdez** [comment]
> > Also, @israel-hdez , if we have reproducibility, what do you think of adding a test case to make sure regression always catches such issues ?
> 
> @rachitchauhan43 I do know how to reproduce, but IMO that should be solved under its own ticket, because we would need to find a way to replicate within CI -- I crashed my dev cluster a few times before I could replicate successfully, so what I did on my device may not be suitable for GitHub Actions. And, of course, we would need to ensure all relevant code paths are covered.

Actually, I think contributors and reviewers should be careful when a `watch` or `list` privilege is being added to the RBAC role ---and this cannot be covered by tests.

This PR is around ConfigMaps, ServiceAccounts and Secrets and is already removing the no longer needed privileges. But any resource that the controller is watching may lead to an OOMKill, if the watch/informer fills the available memory.

**Comment by sivanantha321** [comment]
> Looks like lots of `imports` reorganization happened in this PR. @sivanantha321 : Possible to add as an action itself that fails if imports are not as per standard in `go` files ?

We can do it in a separate PR as many files requires reorganization.

File: cmd/manager/main.go (hunk 9 above)
Comment (sivanantha321) [reply cmd/manager/main.go:116]: done

**Review by israel-hdez — COMMENTED** [review]
Re-approving

/lgtm

**Comment by rachitchauhan43** [comment]
> > Looks like lots of `imports` reorganization happened in this PR. @sivanantha321 : Possible to add as an action itself that fails if imports are not as per standard in `go` files ?
> 
> We can do it in a separate PR as many files requires reorganization.

Sure, I meant that only. Don't want to have multi-purpose PR.

**Comment by rachitchauhan43** [comment]
> > Also, @israel-hdez , if we have reproducibility, what do you think of adding a test case to make sure regression always catches such issues ?
> 
> @rachitchauhan43 I do know how to reproduce, but IMO that should be solved under its own ticket, because we would need to find a way to replicate within CI -- I crashed my dev cluster a few times before I could replicate successfully, so what I did on my device may not be suitable for GitHub Actions. And, of course, we would need to ensure all relevant code paths are covered.

Agreed.  Please create separate ISsue for that.

**Pull request merged**